subprojects/snapshots/src/main/java/org/gradle/internal/vfs/impl/DefaultSnapshotHierarchy.java

```diff
@@ -90,7 +90,7 @@ public SnapshotHierarchy invalidate(String absolutePath, NodeDiffListener diffLi
             return empty();
         }
         return rootNode.invalidate(relativePath, caseSensitivity, diffListener)
-            .<SnapshotHierarchy>map(it -> new DefaultSnapshotHierarchy(it, caseSensitivity))
+            .<SnapshotHierarchy>map(it -> (it == rootNode) ? this : new DefaultSnapshotHierarchy(it, caseSensitivity))
```

Comment:
How does this fare wrt performance? Sounds like something that could potentially have quite a bit of impact on memory pressure, no?

Reply:
We create one small object here for every change. Note that this is only the reference to the root which we recreate, not the whole hierarchy. So this shouldn't have any effect on performance.